### PR TITLE
Platemapper sj

### PR DESCRIPTION
**Pull request by sjanssen2**
here come my first few DB functions

### CODE REVIEW

**Review by qiyunzhu — COMMENTED** [review]
Thanks!

File: knimin/lib/data_access_pm.py [hunk 1]
```diff
@@ -0,0 +1,904 @@
+from knimin.lib.exceptions import *
```
Comment (antgonza) [inline knimin/lib/data_access_pm.py:1]: We normally list the libraries that we are importing. I guess this shows that this repo or files are not automatically being checked with flake8, could you add? What about coveralls? If another PR just adding those and then update this one is easier, we could go with that. Thanks!

File: knimin/lib/data_access_pm.py [hunk 2]
```diff
@@ -0,0 +1,904 @@
...
+
+    Raises
+    ------
+    LabadminDBUnknownIDError
+        If no user with the given email exists.
+    """
+    with TRN:
+        sql = """SELECT email FROM ag.labadmin_users WHERE email = %s"""
+        TRN.add(sql, [email])
+        if len(TRN.execute_fetchindex()) == 0:
```
Comment (antgonza) [inline knimin/lib/data_access_pm.py:26]: I think this can be: if TRN.execute_fetchindex():

File: knimin/lib/data_access_pm.py [hunk 3]
```diff
@@ -0,0 +1,904 @@
...
+        # check that tbl_name exists in DB
+        sql = """SELECT DISTINCT table_name from INFORMATION_SCHEMA.COLUMNS
+                 WHERE table_schema = 'pm' AND table_name = %s"""
+        TRN.add(sql, [tbl_name])
+        if len(TRN.execute_fetchindex()) != 1:
+            raise LabadminDBError('Table %s does not exist in data base.' %
+                                  tbl_name)
+
+        # check that table has exactly three columns
+        sql = """SELECT column_name from INFORMATION_SCHEMA.COLUMNS
```
Comment (antgonza) [inline knimin/lib/data_access_pm.py:64]: Just wondering if the current code is better than:

``` python
sql = """SELECT EXISTS (SELECT 1 from INFORMATION_SCHEMA.COLUMNS WHERE table_schema = 'pm' AND table_name = %s AND column_name in ('name', 'notes', '%s_id'))"""
TRN.add(sql, [tbl_name]*2) 
if TRN.execute_fetchindex():
    raise ...
```

File: knimin/lib/data_access_pm.py [hunk 4]
```diff
@@ -0,0 +1,904 @@
...
+        b) If the table does not follow the expected design, which is that the
+           table must have exactly the three columns:
+           [tbl_name_id, name, notes]
+    """
+    with TRN:
+        # check that tbl_name exists in DB
+        sql = """SELECT DISTINCT table_name from INFORMATION_SCHEMA.COLUMNS
+                 WHERE table_schema = 'pm' AND table_name = %s"""
+        TRN.add(sql, [tbl_name])
+        if len(TRN.execute_fetchindex()) != 1:
```
Comment (antgonza) [inline knimin/lib/data_access_pm.py:59]: The comment about `len(TRN.execute_fetchindex()` applies to all occurrences.

File: knimin/lib/data_access_pm.py [hunk 5]
```diff
@@ -0,0 +1,904 @@
...
+            raise LabadminDBError('Table %s does not exist in data base.' %
+                                  tbl_name)
+
+        # check that table has exactly three columns
+        sql = """SELECT column_name from INFORMATION_SCHEMA.COLUMNS
+                 WHERE table_schema = 'pm' AND table_name = %s"""
+        TRN.add(sql, [tbl_name])
+        column_names = [c[0] for c in TRN.execute_fetchindex()]
```
Comment (antgonza) [inline knimin/lib/data_access_pm.py:67]: You can use: execute_fetchflatten()

File: knimin/lib/data_access_pm.py [hunk 6]
```diff
@@ -0,0 +1,904 @@
...
+        # check that tbl_name exists in DB
+        sql = """SELECT DISTINCT table_name from INFORMATION_SCHEMA.COLUMNS
+                 WHERE table_schema = 'pm' AND table_name = %s"""
+        TRN.add(sql, [tbl_name])
+        if len(TRN.execute_fetchindex()) != 1:
+            raise LabadminDBError('Table %s does not exist in data base.' %
+                                  tbl_name)
+
+        # check that table has the column field
+        sql = """SELECT column_name from INFORMATION_SCHEMA.COLUMNS
```
Comment (antgonza) [inline knimin/lib/data_access_pm.py:113]: Same comment than before, perhaps is easier to do an EXIST and SELECT the table name and field.

File: knimin/lib/data_access_pm.py [hunk 7]
```diff
@@ -0,0 +1,904 @@
...
+                 WHERE table_schema = 'pm' AND table_name = %s"""
+        TRN.add(sql, [tbl_name])
+        if field not in [c[0] for c in TRN.execute_fetchindex()]:
+            raise LabadminDBError(('Table %s does not have the given column: '
+                                   '%s.') % (tbl_name, field))
+
+        # check if an object with the same value already exists
+        sql = """SELECT """+field+""" FROM pm."""+tbl_name+"""
```
Comment (antgonza) [inline knimin/lib/data_access_pm.py:121]: Flake8 will complain about this, one option is:
"SELECT {0} FROM pm.{1} WHERE {0} = %s".format(field, tbl_name)

File: knimin/lib/data_access_pm.py [hunk 8]
```diff
@@ -0,0 +1,904 @@
...
+
+    with TRN:
+        _check_table_layout(tbl_name)
+
+        # check if an object with the same name already exists
+        if _exists(tbl_name, 'name', name):
+            raise LabadminDBDuplicateError(tbl_name, '%s: %s' % ('name', name))
+
+        # add the new object into the DB
+        sql = """INSERT INTO pm."""+tbl_name+""" (name, notes)
```
Comment (antgonza) [inline knimin/lib/data_access_pm.py:174]: Same flake8, suggestion than before. This applies to all occurrences.

**Review by antgonza — COMMENTED** [review]
A few comments. This is a long PR, suggest taking a look at https://github.com/biocore/qiita/blob/master/CONTRIBUTING.md#the-qiita-development-rules. Thanks!

**Pull request closed without merging**